op-conductor/rpc/ws/server_test.go

```diff
@@ -0,0 +1,528 @@
+		t.Error("Timeout waiting for server ping")
+	case <-client.pingsReceived:
+		t.Log("Client received ping from server")
+	}
+}
+
+// TestClientTimeout tests what happens when a client doesn't respond
+func TestClientTimeout(t *testing.T) {
+	_, testHub, server, cleanup := setupTestServer(t)
+	defer cleanup()
+
+	wsURL := "ws" + strings.TrimPrefix(server.URL, "http") + "/ws"
+	ctx, cancel := context.WithTimeout(context.Background(), 15*time.Second)
+	defer cancel()
+
+	// Create client connection but don't process messages
+	conn, resp, err := websocket.Dial(ctx, wsURL, nil)
+	if err != nil {
+		t.Fatalf("Failed to create test client: %v", err)
+	}
+	if resp.StatusCode != http.StatusSwitchingProtocols {
+		t.Fatalf("Expected status code %d, got %d", http.StatusSwitchingProtocols, resp.StatusCode)
+	}
+
+	// Wait for client registration
+	waitForClientCount(t, testHub, 1, 2*time.Second, "Initial connection")
+
+	// Abruptly close the connection to simulate unresponsive client
+	conn.CloseNow()
+
+	// Wait for client unregistration
+	waitForClientCount(t, testHub, 0, 5*time.Second, "Client cleanup after timeout")
+}
+
+// TestMultipleClientsBroadcast tests broadcast functionality with multiple clients
+func TestMultipleClientsBroadcast(t *testing.T) {
+	handler, testHub, server, cleanup := setupTestServer(t)
+	defer cleanup()
+
+	wsURL := "ws" + strings.TrimPrefix(server.URL, "http") + "/ws"
+	ctx, cancel := context.WithTimeout(context.Background(), 10*time.Second)
+	defer cancel()
+
+	numClients := 3
+	var clients []*testClient
+
+	// Connect multiple clients
+	for i := 0; i < numClients; i++ {
+		client, err := newTestClient(ctx, wsURL)
+		if err != nil {
+			t.Fatalf("Failed to create client %d: %v", i, err)
+		}
+		clients = append(clients, client)
+	}
+
+	// Clean up connections
+	defer func() {
+		for _, client := range clients {
+			client.Close()
+		}
+	}()
+
+	// Wait for all clients to be registered
+	waitForClientCount(t, testHub, numClients, 5*time.Second, "All clients connected")
+
+	// Send broadcast messages
+	testMessages := []string{
+		"Hello World!",
+		"Broadcast Test Message 1",
+		"Another test message",
+	}
+
+	for _, msg := range testMessages {
+		handler.BroadcastMessage([]byte(msg))
+	}
+
+	// Verify all clients received all messages
+	for i, client := range clients {
+		receivedMessages := make([]string, 0)
+
+		// Collect messages with timeout
+		msgTimeout := time.After(2 * time.Second)
+		for len(receivedMessages) < len(testMessages) {
+			select {
+			case msg := <-client.messagesReceived:
+				receivedMessages = append(receivedMessages, string(msg))
+			case <-msgTimeout:
+				t.Errorf("Client %d: timeout waiting for messages, got %d/%d", i, len(receivedMessages), len(testMessages))
+				goto nextClient
+			}
+		}
+
+		// Verify message content
+		for j, expected := range testMessages {
+			if j >= len(receivedMessages) || receivedMessages[j] != expected {
```

Comment:
```suggestion
			if receivedMessages[j] != expected {
```

The previous loop ensures to only exit once len(receivedMessages) == len(testMessages). So j will always be less than len(receivedMessages)
- And hypothetically, even if it doesn't, the next line, becomes prone to a panic trying to access `receivedMessages[j]` (for the impossible situation of having j >= len(receivedMessages))